src/item.cpp

```diff
@@ -2776,6 +2776,21 @@ long item::get_property_long( const std::string& prop, long def ) const
 int item::get_quality( const quality_id &id ) const
 {
     int return_quality = INT_MIN;
+
+    /**
+     * EXCEPTION: Items with quality BOIL only count as such if they are empty,
+     * excluding items of their ammo type if they are tools.
+     */
+    if ( id == quality_id( "BOIL" ) && !( contents.empty() || 
+        ( is_tool() && std::all_of( contents.begin(), contents.end(),
+        [this]( const item & itm ) {
+            auto ammo_types = itm.type->ammo->type;
```

Comment:
Looks good except here: `itm.type->ammo` might be a null pointer. `itm.is_ammo` checks for that, but you call it *after* dereferencing the pointer. Also: `ammo_types` does not need to be a copy, a reference would be fine.